README.md
# suncsv

Generates sunrise and sunset calendar items in CSV form for import to your favorite calendar program.

## Setup

First, install the project globally:

```bash
npm install -g suncsv
```

## Usage

To get a CSV with sunrise/sunset events from now until the end of this year:

```bash
suncsv -z YOURZIP > events.csv
```

To get it for next year, you can specify your begin and end dates:

```bash
suncsv -z YOURZIP --begin '2015-01-01' --end '2015-12-31' > events.csv
```

## Add to Google Calendar

Now that you've got your `events.csv` file, you can import it into your favorite calendar program. __Google Calendar__ makes it really easy. We can even create a new calendar just for these items:

* Click on the little down arrow beside __My Calendars__
* Select __Create new calendar__

Once you've created your calendar and are back on the main screen, let's import the events:

* Click on the little down arrow beside __Other calendars__
* Select __Import calendar__
* Click __Choose File__ and find your newly-generated `events.csv`
* Now select the calendar you just created for the new items
* Click __Import__!

That should do it! You've got sunrise and sunset times you can easily turn on and off by toggling your new sub-calendar.

Enjoy!

## License

(The MIT License)

Copyright (c) 2013 Scott Nonnenberg &lt;scott@nonnenberg.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
## 0.2.0 (2021-11-28)

* Better handling of timezones when processing input dates. Now '2022-01-01' means what it should, no matter your timezone.
* Update dependencies and introduce `package-lock.json`

## 0.1.2 (2014-11-21)

* Remove nonexistent `thehelp-time` dependency

## 0.1.1 (2014-09-23)

* Fix name of binary

## 0.1.0 (2014-09-23)

* Initial release


src/server/go.js
'use strict';

// The actual program - the binary juse `require()`s this file.

var path = require('path');

var zip = require('zippity-do-dah');
var commander = require('commander');
var suncalc = require('suncalc');

var util = require('./util');
var config = require(path.join(__dirname, '../../package.json'));

var startup = function startup(today) {
  // Set up some default
  var firstOfNextYear = new Date(today.getFullYear() + 1, 0, 1);
  var endOfThisYear = util.offset(-util.DAY_IN_MIL, firstOfNextYear);

  // Delegate option parsing, help display, etc. to `commander`
  commander
    .version(config.version)
    .option('-z, --zip <zipcode>', 'Zipcode to use for location', '98103')
    .option('-b, --begin <date>', 'Start date (today)', util.makeDate, today)
    .option('-e, --end <date>', 'End date (end of this year)',
      util.makeDate, endOfThisYear)
    .option('-r, --sunrise <boolean>', 'Whether to export sunrise (true)',
      util.makeBoolean, true)
    .option('-s, --sunset <boolean>', 'Whether to export sunset (true)',
      util.makeBoolean, true)
    .parse(process.argv);
};

// Okay, here we go...

var today = new Date();
startup(today);

const options = commander.opts();

var processDay = function processDay(day, gps) {
  // Thanks, `suncalc` for your awesomeness
  var item;
  var times = suncalc.getTimes(day, gps.latitude, gps.longitude);
  var daylight = util.renderTimespan(times.sunrise, times.sunset);

  // Print out sunset and sunrise rows

  if (options.sunrise) {
    item = util.createItem('Sunrise', times.sunrise, daylight);
    console.log(item);
  }

  if (options.sunset) {
    item = util.createItem('Sunset', times.sunset, daylight);
    console.log(item);
  }
};

// Prepare working variables
var gps = zip.zipcode(options.zip);
var day = util.normalizeDate(options.begin, today);
var end = util.normalizeDate(options.end, today);

// Write out the CSV header
console.log(util.csvHeader);

while (day.getTime() <= end.getTime()) {
  processDay(day, gps);

  day = util.offset(util.DAY_IN_MIL, day);
  day = util.normalizeDate(day, today);
}


src/server/util.js

'use strict';

var moment = require('moment');
var duration = require('duration');

exports.DAY_IN_MIL = 1000 * 60 * 60 * 24;

// Input Processing
// ========

// Create a date from a provided string
exports.makeDate = function makeDate(date) {
  return moment(date).toDate();
};

// Create a boolean from a provided string
exports.makeBoolean = function makeBoolean(bool) {
  return (bool !== 'false' && bool !== 'no');
};

// Date Processing
// ========

// Add a given set of milliseconds either to now or to a provided `date`.
exports.offset = function offset(mil, date) {
  date = date || new Date();
  var time = date.getTime();
  return new Date(time + mil);
};

// Ensure that a generated date has the same time as today. Useful for incrementing by
// a day across a daylight savings time boundary.
exports.normalizeDate = function normalizeDate(date, reference) {
  return new Date(
    date.getFullYear(),
    date.getMonth(),
    date.getDate(),
    reference.getHours(),
    reference.getMinutes(),
    reference.getSeconds()
  );
};

// Manually generate the two-column date used in the CSV
exports.formatDate = function formatDate(date) {
  return moment(date).format('YYYY/MM/DD') +
    ',' + moment(date).format('HH:mm:ss');
};

// Render timespan
exports.renderTimespan = function renderTimespan(from, to) {
  return duration(from, to).toString(0, 1);
};

// CSV Output
// ========

/*
Google Calendar CSV docs: <https://support.google.com/calendar/answer/45656>

```
Subject,Start Date,Start Time,End Date,End Time
Final Exam,05/12/20,07:10:00 PM,05/12/20,10:00:00 PM,False
```
*/

// Our events are fifteen minutes long, generally the shortest event length that most
// calendar programs can render well.
exports.getEndTime = function getEndTime(start) {
  return exports.offset(1000 * 50 * 15, start);
};

// CSV header row
exports.csvHeader = 'Subject,Start Date,Start Time,End Date,End Time,Description';

// Generate an entire CSV data row
exports.createItem = function createItem(name, date, daylight) {
  return name + ',' +
    exports.formatDate(date) + ',' +
    exports.formatDate(exports.getEndTime(date)) + ',' +
    'Daylight length: ' + daylight;
};


src/server/index.js

'use strict';

var path = require('path');

var suncalc = require('suncalc');
var zip = require('zippity-do-dah');
var commander = require('commander');

var time = require('thehelp-time');

var config = require(path.join(__dirname, '../../package.json'));

var today = new Date();
var firstOfNextYear = new Date(today.getFullYear() + 1, 0, 1);
var endOfThisYear = time.offset(-time.DAY_IN_MIL, firstOfNextYear);

var makeDate = function makeDate(date) {
  return new Date(date);
};

var makeBoolean = function makeBoolean(bool) {
  return (bool !== 'false' && bool !== 'no');
};

var normalizeDate = function normalizeDate(date) {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(),
    today.getHours(), today.getMinutes(), today.getSeconds());
};

var getEndTime = function getEndTime(start) {
  return time.offset(1000 * 50 * 15, start);
};

var formatDate = function formatDate(date) {
  var result = '';
  result += (date.getMonth() + 1) + '/' + date.getDate() + '/' + date.getFullYear();
  result += ',';
  result += time.addPadding(date.getHours(), 2) +
    ':' + time.addPadding(date.getMinutes(), 2) +
    ':' + time.addPadding(date.getSeconds(), 2);
  return result;
};

var createItem = function createItem(name, date, daylight) {
  var result = name + ',' + formatDate(date) + ',' + formatDate(getEndTime(date));
  result += ',Daylight length: ' + daylight;
  return result;
};

commander
  .version(config.version)
  .option('-z, --zip <zipcode>', 'Zipcode to use for location', '98103')
  .option('-b, --begin <date>', 'Start date (today)', makeDate, today)
  .option('-e, --end <date>', 'End date (end of this year)', makeDate, endOfThisYear)
  .option('-r, --sunrise <boolean>', 'Whether to export sunrise (true)', makeBoolean, true)
  .option('-s, --sunset <boolean>', 'Whether to export sunset (true)', makeBoolean, true)
  .parse(process.argv);

var gps = zip.zipcode(commander.zip);
var day = normalizeDate(commander.begin);
var end = normalizeDate(commander.end);

var total = 0;

// Google Calendar csv docs here: https://support.google.com/calendar/answer/45656
// Subject,Start Date,Start Time,End Date,End Time
// Final Exam,05/12/20,07:10:00 PM,05/12/20,10:00:00 PM,False
console.log('Subject,Start Date,Start Time,End Date,End Time,Description');

while (day.getTime() <= end.getTime()) {
  total += 1;

  var times = suncalc.getTimes(day, gps.latitude, gps.longitude);
  var daylight = times.sunset.getTime() - times.sunrise.getTime();
  daylight = time.renderTimespan(daylight);

  if (commander.sunrise) {
    var item = createItem('Sunrise', times.sunrise, daylight);
    console.log(item);
  }

  if (commander.sunset) {
    var item = createItem('Sunset', times.sunset, daylight);
    console.log(item);
  }

  day = time.offset(time.DAY_IN_MIL, day);
  day = normalizeDate(day);
}
